lib/cksum.c

```diff
@@ -0,0 +1,258 @@
+{
+#ifndef __KERNEL__
+	MD5_CTX context;
+	int     i;
+	int     rc;
+
+	M0_ENTRY();
+
+	M0_PRE(pi != NULL);
+	M0_PRE(curr_context != NULL);
+	M0_PRE(ergo(bvec != NULL && bvec->ov_vec.v_nr != 0,
+		    bvec != NULL && bvec->ov_vec.v_count != NULL &&
+		    bvec->ov_buf != NULL));
+
+	/* This call is for first data unit, need to initialize prev_context */
+	if (flag & M0_PI_CALC_UNIT_ZERO) {
+		pi->pimd5c_hdr.pih_size = sizeof(struct m0_md5_inc_context_pi);
+		rc = MD5_Init((MD5_CTX *)&pi->pimd5c_prev_context);
+		if (rc != 1) {
+			return M0_ERR_INFO(rc, "MD5_Init failed.");
+		}
+	}
+
+	/* memcpy, so that we do not change the prev_context */
+	memcpy(curr_context, &pi->pimd5c_prev_context, sizeof(MD5_CTX));
+
+	/* get the curr context by updating it*/
+	if (bvec != NULL) {
+		for (i = 0; i < bvec->ov_vec.v_nr; i++) {
+			rc = MD5_Update((MD5_CTX *)curr_context,
+					bvec->ov_buf[i],
+					bvec->ov_vec.v_count[i]);
+			if (rc != 1) {
+				return M0_ERR_INFO(rc, "MD5_Update failed."
+						   "curr_context=%p, "
+						   "bvec->ov_buf[%d]=%p, "
+						   "bvec->ov_vec.v_count[%d]=%lu",
+						   curr_context, i,
+						   bvec->ov_buf[i], i,
+						   bvec->ov_vec.v_count[i]);
+			}
+		}
+	}
+
+	/* If caller wants checksum without seed and with seed, caller needs to
+	 * pass 'pi_value_without_seed'. pi_value will be used to return with
+	 * seed checksum. 'pi_value_without_seed' will be used to return non
+	 * seeded checksum.
+	 */
+	if (pi_value_without_seed != NULL) {
+		/* 
+		 * NOTE: MD5_final() changes the context itself and curr_context
+		 * should not be finalised, thus copy it and use it for
+		 * MD5_final
+		 */
+		memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
+
+		rc = MD5_Final(pi_value_without_seed, &context);
+		if (rc != 1) {
+			return M0_ERR_INFO(rc, "MD5_Final failed"
+					   "pi_value_without_seed=%p"
+					   "curr_context=%p",
+					   pi_value_without_seed, curr_context);
+		}
+	}
+
+	/* if seed is passed, memcpy and update the context calculated so far.
+	 * calculate checksum with seed, set the pi_value with seeded checksum.
+	 * If seed is not passed than memcpy context and calculate checksum
+	 * without seed, set the pi_value with unseeded checksum.
+	 * NOTE: curr_context will always have context without seed.
+	 */
+	memcpy((void *)&context, (void *)curr_context, sizeof(MD5_CTX));
+
+	if (seed != NULL) {
+
+		/*
+		 * seed_str have string represention for 3 uint64_t(8 bytes)
+		 * range for uint64_t is 0 to 18,446,744,073,709,551,615 at
+		 * max 20 chars per var, for three var it will be 3*20, +1 '\0'.
+		 * seed_str needs to be 61 bytes, round off and taking 64 bytes.
+		 */
+		char seed_str[64] = {'\0'};
+		snprintf(seed_str, sizeof(seed_str), "%"PRIx64"%"PRIx64"%"PRIx64,
+			 seed->pis_obj_id.f_container, seed->pis_obj_id.f_key,
+			 seed->pis_data_unit_offset);
+		rc = MD5_Update(&context, (unsigned char *)seed_str,
+				sizeof(seed_str));
+		if (rc != 1) {
+
```

Comment:
[* style *] remove unnecessary blank lines. Same at other places in this routine.